src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.netframework.approved.txt

```diff
@@ -238,6 +238,8 @@ namespace NServiceBus
     public class DelayedRetriesSettings : NServiceBus.Configuration.AdvancedExtensibility.ExposeSettings
     {
         public NServiceBus.DelayedRetriesSettings NumberOfRetries(int numberOfRetries) { }
+        public NServiceBus.DelayedRetriesSettings OnMessageBeingRetried(System.Action<NServiceBus.Faults.DelayedRetryMessage> notificationCallback) { }
```

Comment:
What's your train of thought to introduce sync overloads here where most of the core where async makes sense it is just pure async? Is it the fact that you try to smoothen the transition for existing callback APIs? If that is the case given that this is a new API introduced on the corresponding settings is it really such a stretch to let users implement it with `return Task.CompletedTask`?

Reply:
the main reason was that the current notifications API is currently non-async, by introducing an async-only API it will just require adding `Task.CompletedTask` everywhere (which I initially also had to do in all related tests). I considered this to be an unnecessary pain point which we can easily avoid. The sync API doesn't cost us much and it's much more convenient to use. Compared to the handlers API and other APIs, with the overloads here it's very simple to use and will adjust automatically to the user, so there is no increased risk of users going with a sync API where they need async.

Reply:
I'm not against it. I want to discover the larger context or let's say a heuristic when we are going down the route of saying "async only" or we should provide both. So far on the clear IO-bound path, we always went for async only as standard behavior. 

> I considered this to be an unnecessary pain point which we can easily avoid. 

That's an interesting point which brings us back to my heuristics approach. Should we have done that in the past as well and you are now indirectly pushing to have more sync APIs because you believe we went too far with the async only APIs?

> will adjust automatically to the user

Yes luckily now the compiler is able to automatically resolve the right overload. I seem to remember that wasn't the case in previous compiler versions but I could be wrong. 

The reason I'm bringing this up is not to make your life harder. The reason I'm bringing this up is to understand how we would explain the deviation of our standard async-only approach? Or in other words, if someone looks at the code base and asks us "I see you introduced a sync overload for notifications why haven't you done the same for my beloved handlers, mutators etc.". If then our reason is "because it was easy to introduce and we had existing users" I would have a hard time understanding it because for handlers, mutators etc. we also had existing users and then the only point left is "it was easy to introduce".

Or in other words, would we have done the same if you had introduced this change in a major?

Reply:
Since these APIs are kind of niche thing I'd rather have people type `Task.CompletedTask` than make a mistake and use sync version where they should be async.

Reply:
You mean this problem @SzymonPobiega ?

![image](https://user-images.githubusercontent.com/174258/66293693-83edcd00-e8e7-11e9-95e3-37b3b336a8e1.png)

![image](https://user-images.githubusercontent.com/174258/66293728-9405ac80-e8e7-11e9-91a7-0475ada1a811.png)

Reply:
yeah, I'm aware of the deviation from the other, mostly sync-only APIs and in the end, I made a personally biased choice given that I didn't find a clear guideline to go strictly with async APIs.

> Should we have done that in the past as well and you are now indirectly pushing to have more sync APIs because you believe we went too far with the async only APIs?

hard to say regarding existing APIs. I think in case of most APIs we provide, we can't easily provide sync overloads due to the nature of the API (e.g. using the right interface) and the safer version is to go with async only in this case. After all, the async APIs are the must-have requirement (it's what triggered this work after all), everything else is syntactic sugar.

If we can provide an easy & safe way to help the user avoid Task.FromResult/Task.CompletedTask however, then we should do it. But I'm not attached to the sync overloads and I knew this discussion would come up ;) (it's why this is still in draft)

Reply:
I am by no mean as knowledgeable in async/sync as you are @danielmarbach & @timbussmann but anyway I'll say something;)
- I didn't mean any specific problem @danielmarbach, just expressed my observations
- I assume the notification APIs are used by small portion of our customers
- Adding sync variant mean more code in the core that is almost duplicate of async code
- In the past we made a lot of effort to justify our position of not having sync and async variants
- The new APIs are so different from the old ones that the effort required to ass `Task.CompletedTask` is small compared to the change the users need to do to prepare for V8

Reply:
I would agree with @SzymonPobiega especially because existing non-async code can still be in the existing sync APIs. If a user decides to switch over to the new one adding `Task.CompletedTask` is not a big deal and a very well known way of coping with the core's default "async-only" way

Reply:
> Since these APIs are kind of niche thing I'd rather have people type Task.CompletedTask than make a mistake and use sync version where they should be async.

I can't see any risk like that caused by having sync overloads.

I just don't really see how "sticking with async only APIs" makes things better for anyone. It doesn't cost us anything in this case and it saves the user time and ugly code. IMO it's a win-win. The only reason I could see is the potential "inconsistency" on when to provide sync overloads and when not but I think we have a fairly good understand and can clearly explain when it would be too risky and when it's not. Does that mean that we should provide more sync overloads? I don't know but I also don't see a reason why not (again, given the constraints described).